File: rust/year_2018/src/day_01.rs
use itertools::Itertools;

pub fn part_1(input: &str) -> i64 {
    input.lines().map(|line| line.parse::<i64>().unwrap()).sum()
}

pub fn part_2(input: &str) -> i64 {
    Some(0)
        .into_iter()
        .chain(input.lines().map(|line| line.parse().unwrap()).cycle())
        .scan(0, |sum, num| {
            *sum += num;
            Some(*sum)
        })
        .duplicates()
        .next()
        .unwrap_or_else(|| unreachable!())
}

#[cfg(test)]
mod test {
    use test_case::test_case;

    const EXAMPLE_1: &str = "+1\n-2\n+3\n+1";

    const INPUT: &str = include_str!("../test_data/day_01.txt");

    #[test_case(EXAMPLE_1 => 3)]
    #[test_case("+1\n+1\n+1" => 3)]
    #[test_case("+1\n+1\n-2" => 0)]
    #[test_case("-1\n-2\n-3" => -6)]
    #[test_case(INPUT => 599)]
    fn part_1(input: &str) -> i64 {
        super::part_1(input)
    }

    #[test_case(EXAMPLE_1 => 2)]
    #[test_case("+1\n-1" => 0)]
    #[test_case("+3\n+3\n+4\n-2\n-4" => 10)]
    #[test_case("-6\n+3\n+8\n+5\n-6" => 5)]
    #[test_case("+7\n+7\n-2\n-7\n-4" => 14)]
    #[test_case(INPUT => 81204)]
    fn part_2(input: &str) -> i64 {
        super::part_2(input)
    }
}

File: python/year_2018/day_01.py
"""
https://adventofcode.com/2018/day/1
"""

from itertools import accumulate, cycle
from pathlib import Path

from iteration_utilities import duplicates


def part_1(text):
    return sum(map(int, text.split()))


def part_2(text):
    return next(duplicates(accumulate(cycle(map(int, text.split())), initial=0)))


def _read_input() -> str:
    return (Path(__file__).with_name("test_data") / "day_01.txt").read_text()


def test_part_1_input():
    assert part_1(_read_input()) == 599


def test_part_1_example_1():
    assert part_1("+1 -2 +3 +1") == 3


def test_part_1_example_2():
    assert part_1("+1 +1 +1") == 3


def test_part_1_example_3():
    assert part_1("+1 +1 -2") == 0


def test_part_1_example_4():
    assert part_1("-1 -2 -3") == -6


def test_part_2_input():
    assert part_2(_read_input()) == 81204


def test_part_2_example_1():
    assert part_2("+1 -1") == 0


def test_part_2_example_2():
    assert part_2("+3 +3 +4 -2 -4") == 10


def test_part_2_example_3():
    assert part_2("-6 +3 +8 +5 -6") == 5


def test_part_2_example_4():
    assert part_2("+7 +7 -2 -7 -4") == 14
